contracts/FlowCallbackScheduler.cdc

```diff
@@ -1276,19 +1288,30 @@ access(all) contract FlowCallbackScheduler {
                 self.canceledCallbacks.remove(at: 0)
             }
 
-            let callbackHandler = callback.handler.borrow()
-                ?? panic("Invalid callback handler: Could not borrow a reference to the callback handler")
-
-            emit Canceled(
-                id: callback.id,
-                priority: callback.priority.rawValue,
-                feesReturned: refundedFees.balance,
-                feesDeducted: totalFees - refundedFees.balance,
-                callbackOwner: callback.handler.address,
-                callbackHandlerTypeIdentifier: callbackHandler.getType().identifier,
-                callbackName: callbackHandler.name,
-                callbackDescription: callbackHandler.description
-            )
+            if let callbackHandler = callback.handler.borrow() {
+                emit Canceled(
+                    id: callback.id,
+                    priority: callback.priority.rawValue,
+                    feesReturned: refundedFees.balance,
+                    feesDeducted: totalFees - refundedFees.balance,
+                    callbackOwner: callback.handler.address,
+                    callbackHandlerTypeIdentifier: callbackHandler.getType().identifier,
+                    callbackName: callbackHandler.name,
+                    callbackDescription: callbackHandler.description
+                )
+            } else {
+                // if the callback handler is not borrowable, emit a cancel event with empty values
+                emit Canceled(
+                    id: callback.id,
+                    priority: callback.priority.rawValue,
+                    feesReturned: refundedFees.balance,
+                    feesDeducted: totalFees - refundedFees.balance,
+                    callbackOwner: 0x0000000000000000, // invalid address
+                    callbackHandlerTypeIdentifier: "",
+                    callbackName: "",
+                    callbackDescription: ""
```

Comment:
Couldn't `callbackOwner` be retrieved from `callback.handler.address` or `callback.handlerAddress` and `callbackHandlerTypeIdentifier` pulled from `callback.handlerTypeIdentifier`? Is the `0x0` address and empty values a signifier that the capability is invalid?

Reply:
getting the type identifier from just the capability and not the reference isn't that useful because it will just return the generic type, you're right about the address though, so I'll put that back in

Reply:
> getting the type identifier from just the capability and not the reference isn't that useful because it will just return the generic type

But the `CallbackData` (`callback` variable in this scope) struct has a `handlerTypeIdentifier` field which is assigned as the type of the `CallbackHandler` on `init` from the provided Capability, so it should be the concrete type shouldn't it?